File: trunk/symmetric/symmetric-db/src/main/java/org/jumpmind/db/platform/sqlite/SqLitePlatform.java
package org.jumpmind.db.platform.sqlite;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import java.sql.Types;

import javax.sql.DataSource;

import org.jumpmind.db.IDatabasePlatform;
import org.jumpmind.db.platform.AbstractJdbcDatabasePlatform;
import org.jumpmind.util.Log;

/*
 * The platform implementation for the SQLite database.
 */
public class SqLitePlatform extends AbstractJdbcDatabasePlatform implements IDatabasePlatform {

    /* Database name of this platform. */
    public static final String DATABASENAME = "SQLite3";

    /* The standard H2 driver. */
    public static final String JDBC_DRIVER = "org.sqlite.JDBC";

    /*
     * Creates a new instance of the H2 platform.
     */
    public SqLitePlatform(DataSource dataSource, Log log) {
        super(dataSource, log);

        info.setNonPKIdentityColumnsSupported(false);
        info.setIdentityOverrideAllowed(false);
        info.setSystemForeignKeyIndicesAlwaysNonUnique(true);
        info.setNullAsDefaultValueRequired(false);
        info.addNativeTypeMapping(Types.ARRAY, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.DISTINCT, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.NULL, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.REF, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.STRUCT, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.DATALINK, "BINARY", Types.BINARY);

        info.addNativeTypeMapping(Types.BIT, "BOOLEAN", Types.BIT);
        info.addNativeTypeMapping(Types.TINYINT, "SMALLINT", Types.TINYINT);
        info.addNativeTypeMapping(Types.SMALLINT, "SMALLINT", Types.SMALLINT);
        info.addNativeTypeMapping(Types.BINARY, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.BLOB, "BLOB", Types.BLOB);
        info.addNativeTypeMapping(Types.CLOB, "CLOB", Types.CLOB);
        info.addNativeTypeMapping(Types.FLOAT, "DOUBLE", Types.DOUBLE);
        info.addNativeTypeMapping(Types.JAVA_OBJECT, "OTHER");

        info.setDefaultSize(Types.CHAR, Integer.MAX_VALUE);
        info.setDefaultSize(Types.VARCHAR, Integer.MAX_VALUE);
        info.setDefaultSize(Types.BINARY, Integer.MAX_VALUE);
        info.setDefaultSize(Types.VARBINARY, Integer.MAX_VALUE);

        ddlReader = new SqLiteDdlReader(log, this);
        ddlBuilder = new SqLiteBuilder(log, this); 
    }

    public String getName() {
        return DATABASENAME;
    }

}

File: symmetric-client-clib/src/db/platform/sqlite/SqlitePlatform.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "db/platform/sqlite/SqlitePlatform.h"
#include "common/Log.h"

static int SymSqlitePlatform_tableExistsCallback(void *exists, int argc, char **argv, char **columName) {
    *((int *) exists) = argc > 0;
    return 0;
}

int SymSqlitePlatform_tableExists(SymDatabasePlatform *super, char *tableName) {
    int exists = 0;
    char sql[100];
    SymSqlitePlatform *this = (SymSqlitePlatform *) super;
    snprintf(sql, 100, "pragma table_info('%s')", tableName);
    sqlite3_exec(this->db, sql, &SymSqlitePlatform_tableExistsCallback, &exists, NULL);
    return exists;
}

int SymSqlitePlatform_executeSql(SymDatabasePlatform *super,
        const char *sql, int (*callback)(void*, int, char **, char **),
        void *arg, char **errorMessage) {
    SymSqlitePlatform *this = (SymSqlitePlatform *) super;
    return sqlite3_exec(this->db, sql, callback, arg, errorMessage);
}

SymSqliteSqlTemplate * SymSqlitePlatform_getSqlTemplate(SymSqlitePlatform *this) {
    return (SymSqliteSqlTemplate *) this->sqlTemplate;
}

void SymSqlitePlatform_destroy(SymDatabasePlatform *super) {
	SymLog_info("Closing SQLite database");
    SymSqlitePlatform *this = (SymSqlitePlatform *) super;
    sqlite3_close(this->db);
    this->sqlTemplate->destroy(this->sqlTemplate);
//	  free(super->ddlReader);
    free(this);
}

SymSqlitePlatform * SymSqlitePlatform_new(SymSqlitePlatform *this, SymProperties *properties) {
    if (this == NULL) {
        this = (SymSqlitePlatform *) calloc(1, sizeof(SymSqlitePlatform));
    }
    SymDatabasePlatform *super = SymDatabasePlatform_new(&this->super);
    super->databaseInfo.catalogSeparator = ".";
    super->databaseInfo.schemaSeparator = ".";
    super->databaseInfo.delimiterToken = "\"";
    super->ddlReader = (SymDdlReader *) SymSqliteDdlReader_new(NULL, (SymDatabasePlatform *) this);
    super->name = SYM_DATABASE_SQLITE;
    super->version = (char *) sqlite3_libversion();
    super->executeSql = (void *) &SymSqlitePlatform_executeSql;
    super->tableExists = (void *) &SymSqlitePlatform_tableExists;
    super->getSqlTemplate = (void *) &SymSqlitePlatform_getSqlTemplate;
    super->destroy = (void *) &SymSqlitePlatform_destroy;

    SymLog_info("The IDatabasePlatform being used is SymSqlitePlatform");

    char filename[256];
    char *url = properties->get(properties, SYM_PARAMETER_DB_URL, SYM_DATABASE_SQLITE);
    url += strlen(SYM_DATABASE_SQLITE) + 1;
    strncpy(filename, url, 256);
    SymLog_info("Opening SQLite database at %s", filename);

    if (sqlite3_open_v2(filename, &this->db,
            SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_URI, NULL)) {
    	SymLog_error("Can't open database: %s", sqlite3_errmsg(this->db));
        sqlite3_close(this->db);
        return NULL;
    }

    int busyTimeout = atoi(properties->get(properties, SYM_PARAMETER_SQLITE_BUSY_TIMEOUT_MS, SYM_SQLITE_DEFAULT_BUSY_TIMEOUT_MS));

    sqlite3_busy_timeout(this->db, busyTimeout);

    this->sqlTemplate = (SymSqlTemplate *) SymSqliteSqlTemplate_new(NULL, this->db);

    SymLog_info("Detected database '%s', version '%s'", super->name, super->version);

    char *initSql = properties->get(properties, SYM_PARAMETER_SQLITE_INIT_SQL, NULL);
    if (SymStringUtils_isNotBlank(initSql)) {
        int error, i;
        SymLog_debug("Initializing database with '%s'", initSql);
        SymStringArray *array = SymStringArray_split(initSql, ";");
        for (i = 0; i < array->size; i++) {
            char *sql = array->get(array, i);
            if (SymStringUtils_isNotBlank(sql)) {
                this->sqlTemplate->update(this->sqlTemplate, sql, NULL, NULL, &error);
            }
        }
    }

    return this;
}
